src/rust/engine/client/src/pantsd.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use core::fmt;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use libc::pid_t;
+use log::debug;
+use sha2::{Digest, Sha256};
+
+struct Metadata {
+  metadata_dir: PathBuf,
+}
+
+impl Metadata {
+  fn mount<P: AsRef<Path>>(directory: P) -> Result<Metadata, String> {
+    let info = uname::uname().map_err(|e| format!("{}", e))?;
+    let host_hash = Sha256::new()
+      .chain(&info.sysname)
+      .chain(&info.nodename)
+      .chain(&info.release)
+      .chain(&info.version)
+      .chain(&info.machine)
+      .finalize();
+
+    const HOST_FINGERPRINT_LENGTH: usize = 12;
+    let mut hex_digest = String::with_capacity(HOST_FINGERPRINT_LENGTH);
+    for byte in host_hash {
+      fmt::Write::write_fmt(&mut hex_digest, format_args!("{:02x}", byte)).unwrap();
+      if hex_digest.len() >= HOST_FINGERPRINT_LENGTH {
+        break;
+      }
+    }
```

Comment:
This is emulating this:
https://github.com/pantsbuild/pants/blob/8be6fb1d38396dfe0393e5626f85024c4bfb5882/src/python/pants/pantsd/process_manager.py#L72-L88

Reply:
The Python code relies implicitly on the ordering of elements in the return from `os.uname()`. Would it make sense to update the Python code to explicitly refer to the elements invovled? Or maybe comment here that this needs to match the Python?

Also I assume that this Rust code and Python generate identical hashes?

Reply:
They do or else the client would fail to find pantsd metadata (port, pid, etc) and connect.

Reply:
So, the integration test below will fail if the two don't match fwiw.

Reply:
It would be better probably to move this module out into its own crate, expose the method in the `engine` crate as an intrinsic (à la https://github.com/pantsbuild/pants/blob/37f5ddb3081cf3b3b11aa08790e880509fa35f48/src/rust/engine/src/externs/interface.rs#L1829), and then replace the Python implementation. No need to dupe it!

Reply:
Agreed. That I'll save for a follow up in #11960,